### ISSUE TITLE
Review current path resolution scheme

### ISSUE THREAD

**Issue by jpommerening** [LaxarJS/laxar-tooling#5]
This module inherited it's path resolution strategy from `grunt-laxar`:

For each kind of artifact (`flows`, `themes`, `pages`, …) there is expected to be a `laxar-path`-alias defined in the module loader's config. Additionally there is expected to be a `laxar-path-application` and a `laxar-path-default-theme`.

There are two loading schemes: `amd` (just do whatever RequireJS or Webpack does) and `local` (look for the artifact at the associated path). Different artifact types employ different default loading schemes (`amd` for controls, `local` for everything else).

I believe to recall that we agreed to "somehow" move the path configuration out of the module loader into something specifically for our tools (laxar-loader, grunt-laxar) with defaults specified here, in laxar-tooling. Any suggestions regarding to how or where we do that? (e.g.: Should we expect a separate config file? How do we tell laxar-loader where to look for it? etc.)

Should we use the same defaults for all artifacts? Maybe try `local` first, then fall back to what is currently called `amd`?
Should we rename `amd`? Should we keep it around for backwards compatibility and something new?

@alex3683, @jonaswjs, @x1B: thoughts?

**Comment by jpommerening** [comment]
We decided on this:
- `laxar-tooling` provides default paths (see below)
- tooling configuration is expected to be in `laxar.config.js` inside the project's root directory
- keep `amd:` but add `module:` with same behaviour; use `module:` in future projects
- use the same default mechanism for all artifacts: try `local:`, fall back to `module:`

`laxar.config.js` (for current defaults, paths relative to the file):

``` js
module.exports = {
   paths: {
      flows: './application/flows/',
      themes: './includes/themes/',
      pages: './application/pages/',
      layouts: './application/layouts/',
      widgets: './includes/widgets/',
      controls: './includes/controls/',
   // 'default-theme': 'laxar-uikit/themes/default.theme'
   // EDIT: currently, we need a resolvable path
      'default-theme': './bower_components/laxar-uikit/dist/themes/default.theme'
   }
};
```

New defaults (all relative to the project's root):

``` json
{
   "flows": "flows",
   "themes": "themes",
   "pages": "pages",
   "layouts": "layouts",
   "widgets": "widgets",
   "controls": "controls"
}
```

## implementation notes

The new default lookup (with the module fallback) requires little cooperation from the module loader. The `projectPath` callback should return a rejected `Promise` if the requested module does not exist, `laxar-tooling` can then try again with another path. The callback is already asynchronous, so this should not be too complicated. Also, to be in line with `path.resolve`, `loaderContext.resolve` and `require.resolve`, the callback should be called `resolve`.

The `laxar-application` is to be removed: the application root path is the directory the `laxar.config.js` file is in.

RequireJS support should not be completely ignored, but is of secondary importance.

**Comment by x1B** [comment]
One question remaining:

Where is the `laxar-application` root path if there is no `laxar.config.js`? I would suggest the phrasing "the root path is where laxar-tooling _looks for_ the `laxar.config.js`", which probably depends on the invocation (CWD for webpack?).

**Comment by jpommerening** [comment]
Yes, you're right. For webpack I think the [context](http://webpack.github.io/docs/configuration.html#context) option is what we're looking for (plus falling back to `process.cwd()` if it is not set)

**Comment by jpommerening** [comment]
### upgrade info

User-side changes are surprisingly few:
- create a `laxar.config.js` inside the project root directory with the contents above
- if there is no config, the defaults (artifact-type == subdirectory for lookup, e.g. widgets in `widgets/`)
- as of now, `default.theme` _must_ be configured (and be relative to the config file) as it can not easily be deduced
- (optionally) remove the `laxar-path-` mappings from the webpack config

The API (for consumption by [LaxarJS/laxar-loader](/LaxarJS/laxar-loader)) underwent a few changes:
- the `projectPath` callback is now called `resolve` and is supposed to reject the returned promise if the requested module/file does not exist or is not loadable
- the `fileExists` callback, previously used by the asset resolver was dropped in favor of the new behaviour of `projectPath`
- the `log` parameter for creating new artifact collector and asset resolver instances is now an (optional) member of the options object (refer to the API doc for details)

---

An alternative approach to having the webpack/bower/loader-specific `default.theme` path in the `laxar.config.js` would be to keep an alias in the `webpack.config.js`:

`laxar.config.js`:

``` js
module.exports = {
   paths: { 'default-theme': 'default.theme' }
};
```

`webpack.config.js`

``` js
   alias: {
      'default.theme': path.resolve( './includes/lib/laxar-uikit/themes/default.theme' ),
      'polyfills': path.resolve( './includes/lib/laxar/dist/polyfills.js' )
   }
```

**Comment by jpommerening** [comment]
Clarification of the default theme lookup (I think this is different from before):

Resolving a theme named 'default' will trigger a special case in `laxar-tooling` (simplefied, actually paths are `.join()`ed correctly):
- if `paths[ 'default-theme' ]` is not set, try: 
  1. `resolve( paths[ 'themes' ] + '/default.theme' )`  
     (works if there is a default.theme in the themes directory)
  2. `resolve( 'default.theme' )`  
     (works if there's an alias pointing to the default theme)
- if `paths[ 'default-theme' ]` is set:
  1. `resolve( paths[ 'themes' ] + '/' + paths[ 'default-theme' ] )`  
     (works if there is a theme of the configured name in the themes directory)
  2. `resolve( paths[ 'default-theme' ] )`  
     (works if the configured default theme is a "proper" path, or resolvable with webpack)

So, essentially, the default value for `default-theme` is `default.theme`. ¯\_(ツ)_/¯

Tools using `laxar-tooling` are supposed to resolve the paths, relatively to the configuration file, before passing them as options. This results in the third case listed above being no different than the fourth (because `paths[ 'default' ]` would be an absolute path by the time it is `.join()`ed to `paths[ 'themes' ]`).

**Comment by jpommerening** [comment]
Implemented on `master` (`v0.4.0`)
